lib/job-iteration/throttle_enumerator.rb

```diff
@@ -0,0 +1,46 @@
+# typed: true
+# frozen_string_literal: true
+module JobIteration
+  # ThrottleEnumerator allows you to throttle iterations
+  # based on external signal (e.g. database health).
+  # @example
+  #   def build_enumerator(_params, cursor:)
+  #     enumerator_builder.build_throttle_enumerator(
+  #       enumerator_builder.active_record_on_batches(
+  #         Account.inactive,
+  #         cursor: cursor
+  #       ),
+  #       throttle_on: -> { DatabaseStatus.unhealthy? },
+  #       backoff: 30.seconds
+  #     )
+  #   end
+  # The enumerator from above will mimic +active_record_on_batches+,
+  # except when +DatabaseStatus.unhealthy?+ starts to return true.
+  # In that case, it will re-enqueue the job with a specified backoff.
+  class ThrottleEnumerator
+    def initialize(enum, job, throttle_on:, backoff:)
+      @enum = enum
+      @job = job
+      @throttle_on = throttle_on
+      @backoff = backoff
+    end
+
+    def to_enum
+      Enumerator.new(-> { @enum.size }) do |yielder|
+        @enum.each do |*val|
+          if should_throttle?
+            ActiveSupport::Notifications.instrument("throttled.iteration", job_class: @job.class.name)
+            @job.retry_job(wait: @backoff)
+            throw(:abort, :skip_complete_callbacks)
+          end
+
+          yielder.yield(*val)
+        end
+      end
+    end
+
+    def should_throttle?
+      @throttle_on.call
```

Comment:
Is it worth being defensive around this being a proc? What about returning a boolean?

Reply:
Do you mean validating the return of a proc? Or validating that `@throttle_on` is a proc?

Reply:
@moechaieb are. you saying not passing a proc as the `throttle_on` variable?

Reply:
> Do you mean validating the return of a proc? Or validating that @throttle_on is a proc?

I'm asking if it's worth doing both. I'm usually not in favour of overly defensive programming. I think you know best what to do here, and my comment was just a suggestion.

Reply:
It's not always a proc, it can be anything that responds to `call` like:

```ruby
"".method(:upcase)
=> #<Method: String#upcase>
```

We could make it a bit more friendly and raise if the thing doesn't respond to `call`, but I find it pretty unlikely scenario to optimize for.

Reply:
The problem with passing something that is precomputed is that it would be the same value for all the iteration cycle of that job.

Reply:
For a string in my example, yes but for `Database::Status.method(:healthy?)` it kinda doesn't matter.

More over, `throttle_on: Database::Status.method(:healthy?)` vs `throttle_on: -> { Database::Status.healthy? }` saves you an extra stack.